From a93c29058294a706fb2524f5ab076a24c1820c04 Mon Sep 17 00:00:00 2001
From: Giovanni Ricca <giovanniricca@proton.me>
Date: Mon, 23 Mar 2026 12:50:11 +0100
Subject: [PATCH 1/9] Revert "Finish Live Tile on Recents on other displays
 before starting another"

This reverts commit 3e5b2bacf101af2e4af0cbe70ba8a6950f9309a0.

Change-Id: I726aa493a76fdbe3642043a017b6d1e830cddac3
---
 .../quickstep/TaskAnimationManager.java       | 104 +++---------------
 .../quickstep/AbsSwipeUpHandlerTestCase.java  |   9 +-
 .../quickstep/TaskAnimationManagerTest.java   |  98 ++---------------
 .../quickstep/InputConsumerUtilsTest.java     |   7 +-
 4 files changed, 32 insertions(+), 186 deletions(-)

diff --git a/quickstep/src/com/android/quickstep/TaskAnimationManager.java b/quickstep/src/com/android/quickstep/TaskAnimationManager.java
index 1bdd9bb..58d205e 100644
--- a/quickstep/src/com/android/quickstep/TaskAnimationManager.java
+++ b/quickstep/src/com/android/quickstep/TaskAnimationManager.java
@@ -54,7 +54,6 @@ import com.android.quickstep.util.ActiveGestureLog;
 import com.android.quickstep.util.ActiveGestureProtoLogProxy;
 import com.android.quickstep.util.SystemUiFlagUtils;
 import com.android.quickstep.views.RecentsView;
-import com.android.quickstep.views.RecentsViewContainer;
 import com.android.quickstep.window.RecentsWindowManager;
 import com.android.systemui.shared.recents.model.ThumbnailData;
 import com.android.systemui.shared.system.QuickStepContract;
@@ -66,11 +65,8 @@ import dagger.assisted.AssistedFactory;
 import dagger.assisted.AssistedInject;
 
 import java.io.PrintWriter;
-import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Locale;
-import java.util.concurrent.atomic.AtomicInteger;
 
 public class TaskAnimationManager implements RecentsAnimationCallbacks.RecentsAnimationListener {
 
@@ -102,8 +98,6 @@ public class TaskAnimationManager implements RecentsAnimationCallbacks.RecentsAn
             REPOSITORY_INSTANCE = new DaggerSingletonObject<>(
             QuickstepBaseAppComponent::getTaskAnimationManagerRepository);
 
-    private final PerDisplayRepository<TaskAnimationManager> mPerDisplayRepository;
-
     private final TaskStackChangeListener mLiveTileRestartListener = new TaskStackChangeListener() {
         @Override
         public void onActivityRestartAttempt(ActivityManager.RunningTaskInfo task,
@@ -133,12 +127,10 @@ public class TaskAnimationManager implements RecentsAnimationCallbacks.RecentsAn
     public TaskAnimationManager(
             @ApplicationContext Context ctx,
             @Assisted int displayId,
-            DisplayController displayController,
-            PerDisplayRepository<TaskAnimationManager> perDisplayRepository) {
+            DisplayController displayController) {
         mCtx = ctx;
         mDisplayId = displayId;
         mDisplayController = displayController;
-        mPerDisplayRepository = perDisplayRepository;
     }
 
     SystemUiProxy getSystemUiProxy() {
@@ -175,11 +167,6 @@ public class TaskAnimationManager implements RecentsAnimationCallbacks.RecentsAn
                 Log.e(TAG, msg, new Exception());
             }
         }
-
-        final BaseContainerInterface containerInterface = gestureState.getContainerInterface();
-        RecentsAnimationCallbacks newCallbacks = new RecentsAnimationCallbacks(
-                containerInterface.getCreatedContainer());
-
         // Notify if recents animation is still running
         if (mController != null) {
             String msg = "New recents animation started before old animation completed";
@@ -189,84 +176,26 @@ public class TaskAnimationManager implements RecentsAnimationCallbacks.RecentsAn
                 Log.e(TAG, msg, new Exception());
             }
         }
+        // But force-finish it anyways
+        finishRunningRecentsAnimation(
+                /* toHome= */ false,
+                /* forceFinish= */ true,
+                /* forceFinishCb= */ null,
+                /* reason= */ new ActiveGestureLog.CompoundString(
+                        "TaskAnimationManager.startRecentsAnimation: "
+                                + "force finishing previous animation"));
 
-        List<TaskAnimationManager> runningTaskAnimationManagers = new ArrayList<>();
-        mPerDisplayRepository.forEach(false, taskAnimationManager -> {
-            if (taskAnimationManager.isRecentsAnimationRunning()) {
-                runningTaskAnimationManagers.add(taskAnimationManager);
-            }
-        });
-
-        Runnable onRunningRecentsAnimationFinished = () -> startRecentsAnimationInternal(
-                gestureState, intent, listener, newCallbacks);
-
-        AtomicInteger runningAnimationCount = new AtomicInteger(
-                runningTaskAnimationManagers.size());
-        Log.d(TAG, "startRecentsAnimation - running TaskAnimationMnaager count: "
-                + runningAnimationCount.get());
-        if (runningAnimationCount.get() == 0) {
-            onRunningRecentsAnimationFinished.run();
-            return newCallbacks;
-        }
-        for (TaskAnimationManager taskAnimationManager : runningTaskAnimationManagers) {
-            Log.d(TAG, "startRecentsAnimation - finishRunningRecentsAnimation for display "
-                    + taskAnimationManager.mDisplayId);
-            BaseContainerInterface baseContainerInterface;
-            RecentsViewContainer container;
-            RecentsView<?, ?> recentsView;
-            if (taskAnimationManager.mLastGestureState != null
-                    && (baseContainerInterface =
-                    taskAnimationManager.mLastGestureState.getContainerInterface()) != null
-                    && (container = baseContainerInterface.getCreatedContainer()) != null
-                    && (recentsView = container.getOverviewPanel()) != null
-                    && baseContainerInterface.isInLiveTileMode()
-                    && recentsView.getRecentsAnimationController() != null) {
-                recentsView.switchToScreenshot(/* thumbnailDatas= */ null,
-                        /* onFinishRunnable= */ () -> recentsView.finishRecentsAnimation(
-                                /* toHome= */ true,
-                                /* shouldPip= */ false,
-                                /* onFinishComplete= */ () -> {
-                                    Log.d(TAG, "startRecentsAnimation - finished Recents "
-                                            + "animation via RecentsView for display "
-                                            + taskAnimationManager.mDisplayId);
-                                    if (runningAnimationCount.decrementAndGet() == 0) {
-                                        onRunningRecentsAnimationFinished.run();
-                                    }
-                                }));
-            } else {
-                taskAnimationManager.finishRunningRecentsAnimation(
-                        /* toHome= */ false ,
-                        /* forceFinish= */ true,
-                        /* forceFinishCb= */ () -> {
-                            Log.d(TAG, "startRecentsAnimation - finishRunningRecentsAnimation "
-                                    + "for display " + taskAnimationManager.mDisplayId);
-                            if (runningAnimationCount.decrementAndGet() == 0) {
-                                onRunningRecentsAnimationFinished.run();
-                            }
-                        },
-                        /* reason= */ new ActiveGestureLog.CompoundString(
-                                "TaskAnimationManager.startRecentsAnimation: "
-                                        + "force finishing previous animation on display %d",
-                                taskAnimationManager.mDisplayId));
-            }
-
-            if (taskAnimationManager.mCallbacks != null) {
-                // If mCallbacks still != null, that means we are getting this
-                // startRecentsAnimation() before the previous one got
-                // onRecentsAnimationStart(). In that case, cleanup the previous animation so it
-                // doesn't mess up/listen to state changes in this animation.
-                taskAnimationManager.cleanUpRecentsAnimation(taskAnimationManager.mCallbacks);
-            }
+        if (mCallbacks != null) {
+            // If mCallbacks still != null, that means we are getting this startRecentsAnimation()
+            // before the previous one got onRecentsAnimationStart(). In that case, cleanup the
+            // previous animation so it doesn't mess up/listen to state changes in this animation.
+            cleanUpRecentsAnimation(mCallbacks);
         }
-        return newCallbacks;
-    }
 
-    private void startRecentsAnimationInternal(
-            @NonNull GestureState gestureState, Intent intent,
-            RecentsAnimationCallbacks.RecentsAnimationListener listener,
-            RecentsAnimationCallbacks newCallbacks) {
         final BaseContainerInterface containerInterface = gestureState.getContainerInterface();
         mLastGestureState = gestureState;
+        RecentsAnimationCallbacks newCallbacks = new RecentsAnimationCallbacks(
+                containerInterface.getCreatedContainer());
         mCallbacks = newCallbacks;
         mLauncherDestroyCallbackSet = false;
         final Runnable recentsAnimationStartTimeoutCallback = () -> {
@@ -467,6 +396,7 @@ public class TaskAnimationManager implements RecentsAnimationCallbacks.RecentsAn
         ActiveGestureProtoLogProxy.logSettingRecentsAnimationStartPending(
                 mRecentsAnimationStartPending);
         gestureState.setState(STATE_RECENTS_ANIMATION_INITIALIZED);
+        return mCallbacks;
     }
 
     /**
diff --git a/quickstep/tests/multivalentTests/src/com/android/quickstep/AbsSwipeUpHandlerTestCase.java b/quickstep/tests/multivalentTests/src/com/android/quickstep/AbsSwipeUpHandlerTestCase.java
index f626495..16b741a 100644
--- a/quickstep/tests/multivalentTests/src/com/android/quickstep/AbsSwipeUpHandlerTestCase.java
+++ b/quickstep/tests/multivalentTests/src/com/android/quickstep/AbsSwipeUpHandlerTestCase.java
@@ -70,7 +70,6 @@ import android.view.ViewTreeObserver;
 import androidx.annotation.NonNull;
 import androidx.test.platform.app.InstrumentationRegistry;
 
-import com.android.app.displaylib.fakes.FakePerDisplayRepository;
 import com.android.launcher3.DeviceProfile;
 import com.android.launcher3.LauncherRootView;
 import com.android.launcher3.anim.AnimationSuccessListener;
@@ -280,12 +279,8 @@ public abstract class AbsSwipeUpHandlerTestCase<
     @Before
     public void setUpRecentsContainer() {
         DisplayController displayController = DisplayController.INSTANCE.get(mContext);
-        FakePerDisplayRepository<TaskAnimationManager> fakePerDisplayRepository =
-                new FakePerDisplayRepository<>();
-        TaskAnimationManager taskAnimationManager = new TaskAnimationManager(mContext, mDisplayId,
-                displayController, fakePerDisplayRepository);
-        fakePerDisplayRepository.add(mDisplayId, taskAnimationManager);
-        mTaskAnimationManager = spy(taskAnimationManager);
+        mTaskAnimationManager = spy(
+                new TaskAnimationManager(mContext, mDisplayId, displayController));
         RECENTS_CONTAINER recentsContainer = getRecentsContainer();
         RECENTS_VIEW recentsView = getRecentsView();
 
diff --git a/quickstep/tests/multivalentTests/src/com/android/quickstep/TaskAnimationManagerTest.java b/quickstep/tests/multivalentTests/src/com/android/quickstep/TaskAnimationManagerTest.java
index ffbdaa6..135c808 100644
--- a/quickstep/tests/multivalentTests/src/com/android/quickstep/TaskAnimationManagerTest.java
+++ b/quickstep/tests/multivalentTests/src/com/android/quickstep/TaskAnimationManagerTest.java
@@ -16,8 +16,6 @@
 
 package com.android.quickstep;
 
-import static android.view.Display.DEFAULT_DISPLAY;
-
 import static com.android.quickstep.TaskAnimationManager.RECENTS_ANIMATION_START_TIMEOUT_MS;
 
 import static org.junit.Assert.assertEquals;
@@ -29,12 +27,8 @@ import static org.mockito.ArgumentMatchers.anyInt;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.spy;
-import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
-import static org.mockito.kotlin.OngoingStubbingKt.whenever;
 
 import android.app.ActivityManager;
 import android.app.ActivityOptions;
@@ -44,6 +38,7 @@ import android.content.res.Configuration;
 import android.graphics.Rect;
 import android.os.Bundle;
 import android.os.SystemClock;
+import android.view.Display;
 import android.view.RemoteAnimationTarget;
 import android.view.SurfaceControl;
 import android.window.TransitionInfo;
@@ -52,11 +47,8 @@ import androidx.test.ext.junit.runners.AndroidJUnit4;
 import androidx.test.filters.SmallTest;
 import androidx.test.platform.app.InstrumentationRegistry;
 
-import com.android.app.displaylib.fakes.FakePerDisplayRepository;
-import com.android.launcher3.uioverrides.QuickstepLauncher;
 import com.android.launcher3.util.DisplayController;
 import com.android.launcher3.util.Executors;
-import com.android.quickstep.views.RecentsView;
 import com.android.systemui.shared.system.RecentsAnimationControllerCompat;
 
 import org.junit.Before;
@@ -84,22 +76,25 @@ public class TaskAnimationManagerTest {
     private SystemUiProxy mSystemUiProxy;
 
     private TaskAnimationManager mTaskAnimationManager;
-    @Mock
-    private TaskAnimationManager mTaskAnimationManagerForExternalDisplay;
-    private FakePerDisplayRepository<TaskAnimationManager> mPerDisplayRepository;
+    private TaskAnimationManager mTaskAnimationManagerWithExternalDisplay;
 
     @Before
     public void setUp() {
         DisplayController displayController = DisplayController.INSTANCE.get(mContext);
-        mPerDisplayRepository = new FakePerDisplayRepository<>();
-        mTaskAnimationManager = spy(new TaskAnimationManager(mContext, DEFAULT_DISPLAY,
-                displayController, mPerDisplayRepository) {
+        mTaskAnimationManager = new TaskAnimationManager(mContext, Display.DEFAULT_DISPLAY,
+                displayController) {
             @Override
             SystemUiProxy getSystemUiProxy() {
                 return mSystemUiProxy;
             }
-        });
-        mPerDisplayRepository.add(DEFAULT_DISPLAY, mTaskAnimationManager);
+        };
+        mTaskAnimationManagerWithExternalDisplay =
+                new TaskAnimationManager(mContext, EXTERNAL_DISPLAY_ID, displayController) {
+                    @Override
+                    SystemUiProxy getSystemUiProxy() {
+                        return mSystemUiProxy;
+                    }
+                };
     }
 
     @Test
@@ -120,75 +115,6 @@ public class TaskAnimationManagerTest {
                 optionsCaptor.getValue().getPendingIntentBackgroundActivityStartMode());
     }
 
-    @Test
-    public void startRecentsActivity_finishRecentsBeforeStartingAnother() {
-        final LauncherActivityInterface activityInterface = mock(LauncherActivityInterface.class);
-        final GestureState gestureState = mock(GestureState.class);
-        final RecentsAnimationCallbacks.RecentsAnimationListener listener =
-                mock(RecentsAnimationCallbacks.RecentsAnimationListener.class);
-        doReturn(activityInterface).when(gestureState).getContainerInterface();
-        mPerDisplayRepository.add(EXTERNAL_DISPLAY_ID, mTaskAnimationManagerForExternalDisplay);
-        when(mTaskAnimationManager.isRecentsAnimationRunning()).thenReturn(true);
-        runOnMainSync(() ->
-                mTaskAnimationManager.startRecentsAnimation(gestureState, new Intent(), listener));
-        ArgumentCaptor<Runnable> finishCallback = ArgumentCaptor.forClass(Runnable.class);
-        verify(mTaskAnimationManager, times(1))
-                .finishRunningRecentsAnimation(anyBoolean(), anyBoolean(), finishCallback.capture(),
-                        any());
-        runOnMainSync(() -> finishCallback.getValue().run());
-        verify(mSystemUiProxy).startRecentsTransition(any(), any(), any(), anyBoolean(),
-                any(), anyInt());
-        verify(mTaskAnimationManagerForExternalDisplay, never())
-                .finishRunningRecentsAnimation(anyBoolean(), anyBoolean(), any(), any());
-    }
-
-    @Test
-    public void startRecentsActivity_finishAllRecentsBeforeStartingAnother() {
-        final LauncherActivityInterface activityInterface = mock(LauncherActivityInterface.class);
-        final GestureState gestureState = mock(GestureState.class);
-        final RecentsAnimationCallbacks.RecentsAnimationListener listener =
-                mock(RecentsAnimationCallbacks.RecentsAnimationListener.class);
-        doReturn(activityInterface).when(gestureState).getContainerInterface();
-        RecentsView<?, ?> recentsView = mock(RecentsView.class);
-        QuickstepLauncher quickstepLauncher = mock(QuickstepLauncher.class);
-        doReturn(quickstepLauncher).when(activityInterface).getCreatedContainer();
-        doReturn(recentsView).when(quickstepLauncher).getOverviewPanel();
-        whenever(recentsView.getRecentsAnimationController()).thenReturn(
-                mock(RecentsAnimationController.class));
-        whenever(activityInterface.isInLiveTileMode()).thenReturn(true);
-
-        // Start once on default display to make sure mLastGestureState is set.
-        runOnMainSync(() ->
-                mTaskAnimationManager.startRecentsAnimation(gestureState, new Intent(), listener));
-        verify(mSystemUiProxy).startRecentsTransition(any(), any(), any(), anyBoolean(),
-                any(), eq(DEFAULT_DISPLAY));
-
-        mPerDisplayRepository.add(EXTERNAL_DISPLAY_ID, mTaskAnimationManagerForExternalDisplay);
-        whenever(mTaskAnimationManagerForExternalDisplay.isRecentsAnimationRunning()).thenReturn(
-                true);
-        whenever(mTaskAnimationManager.isRecentsAnimationRunning()).thenReturn(
-                true);
-
-        runOnMainSync(() ->
-                mTaskAnimationManager.startRecentsAnimation(gestureState, new Intent(), listener));
-        ArgumentCaptor<Runnable> screenshotFinishCaptor = ArgumentCaptor.forClass(Runnable.class);
-        verify(recentsView, times(1)).switchToScreenshot(any(), screenshotFinishCaptor.capture());
-        screenshotFinishCaptor.getValue().run();
-        ArgumentCaptor<Runnable> finishRecentsFinishCaptor = ArgumentCaptor.forClass(
-                Runnable.class);
-        verify(recentsView, times(1)).finishRecentsAnimation(anyBoolean(), anyBoolean(),
-                finishRecentsFinishCaptor.capture());
-        finishRecentsFinishCaptor.getValue().run();
-        ArgumentCaptor<Runnable> finishCallback = ArgumentCaptor.forClass(Runnable.class);
-        verify(mTaskAnimationManagerForExternalDisplay, times(1))
-                .finishRunningRecentsAnimation(anyBoolean(), anyBoolean(), finishCallback.capture(),
-                        any());
-
-        runOnMainSync(() -> finishCallback.getValue().run());
-        verify(mSystemUiProxy, times(2)).startRecentsTransition(any(), any(), any(), anyBoolean(),
-                any(), eq(DEFAULT_DISPLAY));
-    }
-
     @Test
     public void testLauncherDestroyed_whileRecentsAnimationStartPending_finishesAnimation() {
         final GestureState gestureState = buildMockGestureState();
diff --git a/quickstep/tests/src/com/android/quickstep/InputConsumerUtilsTest.java b/quickstep/tests/src/com/android/quickstep/InputConsumerUtilsTest.java
index 835a4a6..c75fdff 100644
--- a/quickstep/tests/src/com/android/quickstep/InputConsumerUtilsTest.java
+++ b/quickstep/tests/src/com/android/quickstep/InputConsumerUtilsTest.java
@@ -44,7 +44,6 @@ import androidx.test.ext.junit.runners.AndroidJUnit4;
 import androidx.test.filters.SmallTest;
 import androidx.test.platform.app.InstrumentationRegistry;
 
-import com.android.app.displaylib.fakes.FakePerDisplayRepository;
 import com.android.launcher3.DeviceProfile;
 import com.android.launcher3.R;
 import com.android.launcher3.anim.AnimatedFloat;
@@ -142,11 +141,7 @@ public class InputConsumerUtilsTest {
     @Before
     public void setupTaskAnimationManager() {
         DisplayController displayController = DisplayController.INSTANCE.get(mContext);
-        FakePerDisplayRepository<TaskAnimationManager> fakePerDisplayRepository =
-                new FakePerDisplayRepository<>();
-        mTaskAnimationManager = new TaskAnimationManager(mContext, mDisplayId, displayController,
-                fakePerDisplayRepository);
-        fakePerDisplayRepository.add(mDisplayId, mTaskAnimationManager);
+        mTaskAnimationManager = new TaskAnimationManager(mContext, mDisplayId, displayController);
     }
 
     @Before
-- 
2.53.0

